File: core/src/com/desitum/library/game_objects/GameObject.java
package com.desitum.library.game_objects;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector3;
import com.desitum.library.animation.Animator;
import com.desitum.library.animation.MovementAnimator;

import java.util.ArrayList;

/**
 * Created by kody on 12/27/15.
 * can be used by kody and people in [kody}]
 */
@SuppressWarnings("unused")
public class GameObject extends Sprite implements Comparable<GameObject> {

    public static final int DEFAULT_Z = 0;
    private ArrayList<Animator> mAnimators;
    private ArrayList<Animator> mAnimatorsToRemove;
    private OnFinishedMovingListener mOnFinishedMovingListener;
    private int mZ;

    private float mSpeed, mSpeedX, mSpeedY, mGravityX, mGravityY, mRotationSpeed, mRotationResistance;
    private float[] moveTo;

    public GameObject(TextureRegion textureRegion) {
        super(textureRegion);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(Texture texture) {
        super(texture);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(TextureRegion textureRegion, float width, float height) {
        super(textureRegion);
        setSize(width, height);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(Texture texture, float width, float height) {
        super(texture);
        setSize(width, height);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(TextureRegion textureRegion, float width, float height, float x, float y) {
        super(textureRegion);
        setSize(width, height);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(Texture texture, float width, float height, float x, float y) {
        super(texture);
        setSize(width, height);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        mZ = DEFAULT_Z;
    }

    public GameObject(Texture texture, float width, float height, float x, float y, int z) {
        super(texture);
        setSize(width, height);
        mAnimators = new ArrayList<Animator>();
        mAnimatorsToRemove = new ArrayList<Animator>();
        this.mZ = z;
    }

    public void update(float delta) {
        updateAnimators(delta);
        updatePosition(delta);
    }

    private void updatePosition(float delta) {
        setSpeedX(getSpeedX() + getGravityX() * delta);
        setSpeedY(getSpeedY() + getGravityY() * delta);
        setX(getX() + getSpeedX());
        setY(getY() + getSpeedY());
        setRotationSpeed(getRotationSpeed() * mRotationResistance);
        if (moveTo != null) {
            updateMovement();
        }
    }

    private void updateAnimators(float delta) {
        if (!mAnimators.isEmpty()) {
            for (Animator animator: mAnimators) {
                animator.update(delta);
                if (animator.didFinish()) {
                    mAnimatorsToRemove.add(animator);
                }
            }
            for (Animator animator: mAnimatorsToRemove) {
                mAnimators.remove(animator);
            }
            mAnimatorsToRemove.clear();
        }
    }

    private void updateMovement() {
        if (moveTo != null) {
            if (mSpeedX < 0) {
                if (getX() < moveTo[0]) {
                    moveFinished();
                }
            } else if (mSpeedX > 0) {
                if (getX() > moveTo[0]) {
                    moveFinished();
                }
            } if (mSpeedY < 0) {
                if (getY() < moveTo[1]) {
                    moveFinished();
                }
            } else if (mSpeedY > 0) {
                if (getY() > moveTo[1]) {
                    moveFinished();
                }
            }
        }
    }

    private void moveFinished() {
        setX(moveTo[0]);
        setY(moveTo[1]);
        moveTo = null;
        if (mOnFinishedMovingListener != null) {
            mOnFinishedMovingListener.onFinished(this);
        }
    }

    @SuppressWarnings("unused")
    public void moveToPosition(float x, float y) {
        // x = cos
        // y = sin
        float deltaX = getX() - x;
        float deltaY = getY() - y;
        float angle = (float) Math.atan2(deltaY, deltaX); // in radians
        mSpeedX = (float) Math.cos(angle) * mSpeed;
        mSpeedY = (float) Math.sin(angle) * mSpeed;
    }

    /**
     * Called when user touches the {@link GameObject}
     * @param touchPos position of touch in {@link com.badlogic.gdx.graphics.Camera} coordinates
     * @param touchDown whether the touch was down or not (touchDown, touchUp)
     */
    @SuppressWarnings("unused")
    public void updateTouchInput(Vector3 touchPos, boolean touchDown) {

    }

    @SuppressWarnings("unused")
    public void smoothMoveTo(float x, float y, float duration, int interpolation) {
        MovementAnimator xAnimator = new MovementAnimator(this, getX(), x, duration, 0, interpolation, true, false);
        MovementAnimator yAnimator = new MovementAnimator(this, getY(), y, duration, 0, interpolation, false, true);
        xAnimator.start(true);
        yAnimator.start(true);
        this.mAnimators.add(xAnimator);
        this.mAnimators.add(yAnimator);
    }

    @SuppressWarnings("unused")
    public void addAndStartAnimator(Animator anim) {
        anim.setSprite(this);
        mAnimators.add(anim);
    }

    public int getZ() {
        return mZ;
    }

    @Override
    public int compareTo(GameObject gameObject2D) {
        return this.getZ() - gameObject2D.getZ();
    }

    @SuppressWarnings("unused")
    public void setZ(int z) {
        this.mZ = z;
    }

    @SuppressWarnings("unused")
    public float getSpeed() {
        return mSpeed;
    }

    @SuppressWarnings("unused")
    public void setSpeed(float speed) {
        this.mSpeed = speed;
    }

    @SuppressWarnings("unused")
    public float getSpeedX() {
        return mSpeedX;
    }

    @SuppressWarnings("unused")
    public void setSpeedX(float speedX) {
        this.mSpeedX = speedX;
    }

    @SuppressWarnings("unused")
    public float getSpeedY() {
        return mSpeedY;
    }

    @SuppressWarnings("unused")
    public void setSpeedY(float speedY) {
        this.mSpeedY = speedY;
    }

    @SuppressWarnings("unused")
    public float getGravityX() {
        return mGravityX;
    }

    @SuppressWarnings("unused")
    public void setGravityX(float gravityX) {
        this.mGravityX = gravityX;
    }

    @SuppressWarnings("unused")
    public float getGravityY() {
        return mGravityY;
    }

    @SuppressWarnings("unused")
    public void setGravityY(float gravityY) {
        this.mGravityY = gravityY;
    }

    @SuppressWarnings("unused")
    public float getRotationSpeed() {
        return mRotationSpeed;
    }

    @SuppressWarnings("unused")
    public void setRotationSpeed(float rotationSpeed) {
        this.mRotationSpeed = rotationSpeed;
    }

    @SuppressWarnings("unused")
    public float getRotationResistance() {
        return mRotationResistance;
    }

    /**
     * Set rotation resistance where less than 1 slows down
     * more than 1 mSpeed up rotation
     * and where 0 stops immediately
     * @param rotationResistance rate to slow down by
     */
    @SuppressWarnings("unused")
    public void setRotationResistance(float rotationResistance) {
        this.mRotationResistance = rotationResistance;
    }

    @SuppressWarnings("unused")
    public void setOnFinishedMovingListener(OnFinishedMovingListener onFinishedMovingListener) {
        this.mOnFinishedMovingListener = onFinishedMovingListener;
    }

    @SuppressWarnings("unused")
    public OnFinishedMovingListener getOnFinishedMovingListener() {
        return mOnFinishedMovingListener;
    }

    public void dispose() {
        try {
            getTexture().dispose();
        } catch (Exception e) {
            // Texture has been disposed of elsewhere
        }
    }
}

File: core/src/com/desitum/library/game_objects/GameObject.kt
package com.desitum.library.game_objects

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.utils.viewport.Viewport
import com.desitum.library.animation.Animator
import com.desitum.library.animation.MovementAnimator
import com.desitum.library.drawing.Drawable
import com.desitum.library.game.G2DSprite
import com.desitum.library.game.World
import com.desitum.library.view.TouchEvent
import java.util.*

/**
 * Created by kody on 12/27/15.
 * can be used by kody and people in [kody}]
 */
open class GameObject(textureRegion: TextureRegion?, world: World?) : G2DSprite() {
    var animators: ArrayList<Animator>
        private set
    var onFinishedMovingListener: OnFinishedMovingListener? = null

    var speed: Float = 0f
    private var speedX: Float = 0f
    private var speedY: Float = 0f
    var gravityX: Float = 0f
    var gravityY: Float = 0f
    var rotationSpeed: Float = 0f
    var rotationResistance: Float = 0f
    var focus: Boolean = false
    var focusable = false
        @JvmName("isFocusable") get
    var drawable: Drawable? = null
    private var moveTo: FloatArray? = null
    private var world: World? = null

    init {
        this.world = world
        textureRegion?.let { setRegion(it) }
    }

    override fun update(delta: Float) {
        updateAnimators(delta)
        updatePosition(delta)
    }

    private fun updatePosition(delta: Float) {
        speedX += gravityX * delta
        speedY += gravityY * delta
        x += speedX
        y += speedY
        rotationSpeed *= rotationResistance
        if (moveTo != null) {
            updateMovement()
        }
    }

    private fun updateAnimators(delta: Float) {
        if (!animators.isEmpty()) {
            animators.forEach { animator -> animator.update(delta) }
            animators.removeIf { animator -> animator.didFinish() }
        }
    }

    private fun updateMovement() {
        if (speedX < 0) {
            if (x < moveTo!![0]) {
                moveFinished()
            }
        } else if (speedX > 0) {
            if (x > moveTo!![0]) {
                moveFinished()
            }
        }
        if (speedY < 0) {
            if (y < moveTo!![1]) {
                moveFinished()
            }
        } else if (speedY > 0) {
            if (y > moveTo!![1]) {
                moveFinished()
            }
        }
    }

    private fun moveFinished() {
        x = moveTo!![0]
        y = moveTo!![1]
        moveTo = null
        onFinishedMovingListener?.onFinished(this)
    }

    fun moveToPosition(x: Float, y: Float) {
        // x = cos
        // y = sin
        val deltaX = getX() - x
        val deltaY = getY() - y
        val angle = Math.atan2(deltaY.toDouble(), deltaX.toDouble()).toFloat() // in radians
        speedX = Math.cos(angle.toDouble()).toFloat() * speed
        speedY = Math.sin(angle.toDouble()).toFloat() * speed
        moveTo = floatArrayOf(x, y)
    }

    fun smoothMoveTo(x: Float, y: Float, duration: Float, interpolation: Int) {
        val xAnimator = MovementAnimator(this, getX(), x, duration, 0f, interpolation, true, false)
        val yAnimator = MovementAnimator(this, getY(), y, duration, 0f, interpolation, false, true)
        xAnimator.start(true)
        yAnimator.start(true)
        this.animators.add(xAnimator)
        this.animators.add(yAnimator)
    }

    fun addAndStartAnimator(anim: Animator) {
        anim.sprite = this
        animators.add(anim)
        anim.start(false)
    }

    fun addAnimator(anim: Animator) {
        anim.sprite = this
        animators.add(anim)
    }

    fun dispose() {
        try {
            texture.dispose()
        } catch (e: Exception) {
            // Texture has been disposed of elsewhere
        }
    }

    // -----------------------------
    // region Input methods
    // -----------------------------
    fun dispatchTouchEvent(touchEvent: TouchEvent): Boolean {
        if (touchEvent.action === TouchEvent.Action.DOWN && focusable) {
            world?.requestFocus(this)
        }
        return onTouchEvent(touchEvent)
    }

    open fun onTouchEvent(touchEvent: TouchEvent): Boolean {
        return true
    }

    fun isTouching(touchEvent: TouchEvent): Boolean {
        return boundingRectangle.contains(touchEvent.x, touchEvent.y)
    }
    // -----------------------------
    // endregion
    // -----------------------------

//    // -----------------------------
//    // region Drawing methods
//    // -----------------------------
    fun draw(batch: Batch, viewport: Viewport) {
        if (texture != null) draw(batch)
        /*
        Need to draw in the following order

        1. Background
        2. Draw view itself
        3. Draw any possible children
        4. Draw foreground
         */
        // 1. Draw background
        drawable?.draw(batch, x, y, width, height)

        // 2. Draw view itself
        onDraw(batch, viewport)

        // 3. Draw any possible children
        dispatchDraw(batch, viewport)

        // 4. Draw the foreground / view decorations
        drawForeground(batch, viewport)
    }

    open fun onDraw(spriteBatch: Batch, viewport: Viewport) {

    }

    fun dispatchDraw(spriteBatch: Batch, viewport: Viewport) {

    }

    fun drawForeground(spriteBatch: Batch, viewport: Viewport) {

    }
//    // -----------------------------
//    // endregion
//    // -----------------------------

    fun clearFocus() {
        focus = false
    }

    companion object {
        val DEFAULT_Z = 0
    }

    init {
        animators = ArrayList<Animator>()
        z = DEFAULT_Z
    }
}
